src/tunnel.cc

```diff
@@ -993,40 +1000,56 @@ tunnelErrorComplete(int fd/*const Comm::ConnectionPointer &*/, void *data, size_
         tunnelState->server.conn->close();
 }
 
+bool
+TunnelStateData::retry()
+{
+    {
+        const Comm::Connection &failedDest = *serverDestinations.front();
+        debugs(26, 4, "removing the failed one from " << serverDestinations.size() <<
+                " destinations: " << failedDest);
+    }
+
+    serverDestinations.erase(serverDestinations.begin());
+
+    if (!serverDestinations.empty() && FwdState::EnoughTimeToReForward(startTime)) {
+        debugs(26, 4, "re-forwarding");
+        startConnecting();
+        return true;
+    }
+    return false;
+}
+
+void
+TunnelStateData::sendError(const int xerrno, ErrorState *error, const Comm::ConnectionPointer &conn)
+{
+    assert(error);
+    assert(conn);
+
+    debugs(26, 4, "terminate with error.");
+
+    *status_ptr = error->httpStatus;
+    error->xerrno = xerrno;
+    // on timeout is this still:    error->xerrno = ETIMEDOUT;
+    error->port = conn->remote.port();
+    error->callback = tunnelErrorComplete;
+    error->callback_data = this;
+    errorSend(client.conn, error);
+    if (request)
+        request->hier.stopPeerClock(false);
```

Comment:
The old code did not do this on TLS handshake failures (only on TCP connection failures). My understanding is that we should stop this clock in both cases.

Reply:
* The clock should be started (or resumed) when we are using peers (for any reason).
* The clock should be stopped (or paused) when we are not using peers (for any reason).

I did not check the code, but it feels like a good idea to stop/pause the clock after any peer error (including TLS error), as you suggest. If this is the last peer we can try (_right now_), then we do not need to worry about stopping the clock some time later, when we discover that there are no more peers we can immediately try. In v5/master, there can be a delay between previous peer usage end and next peer usage start...

We should not forget to resume the clock when we _start_ using a peer. I trust there is already code for that.

Reply:
> In v5/master, there can be a delay between previous peer usage end and next peer usage start...

... However, I do not know whether this clock should be paused while Squid is finding new peers to use. If it is supposed to reflect all time spent on Squid-server/peer work, then it should not be paused. The same logic applies to TCP-level errors though so we can keep both code snippets the same.